frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -2448,19 +2472,51 @@ const IR::Node* TypeInference::postorder(IR::Member* expression) {
                 return expression;
             }
         }
-        if (type->is<IR::Type_Header>()) {
-            if (member == IR::Type_Header::setValid ||
-                member == IR::Type_Header::setInvalid) {
+        if (type->is<IR::Type_Header>() || type->is<IR::Type_HeaderUnion>()) {
+            if (type->is<IR::Type_HeaderUnion>() &&
+                (member == IR::Type_Header::setValid ||
+                member == IR::Type_Header::setInvalid)) {
+                ::error(ErrorType::ERR_INVALID, "setValid and setInvalid are "
+                        "not defined for header unions", expression);
+                return expression;
+            }
+            // Built-in methods
+            if ((member == IR::Type_Header::setValid ||
+                member == IR::Type_Header::setInvalid) &&
+                type->is<IR::Type_Header>()) {
                 if (!isLeftValue(expression->expr))
                     typeError("%1%: must be applied to a left-value", expression);
-                // Built-in method
                 auto type = new IR::Type_Method(IR::Type_Void::get(), new IR::ParameterList);
                 auto ctype = canonicalize(type);
                 if (ctype == nullptr)
                     return expression;
                 setType(getOriginal(), ctype);
                 setType(expression, ctype);
                 return expression;
+            } else if ((member == IR::Type_Header::sizeBits) ||
+                       (member == IR::Type_Header::sizeBytes)) {
+                int sz = 0;
+                auto t = typeMap->getType(expression->expr, true);
+                const IR::Type_StructLike* ht;
+                if (t->is<IR::Type_Header>()) {
+                    ht = t->to<IR::Type_Header>();
+                } else if (t->is<IR::Type_HeaderUnion>()) {
+                    ht = t->to<IR::Type_HeaderUnion>();
+                }
+                if (ht == nullptr) {
+                    ::error("%1%: null header/headerUnion?", expression);
+                    return expression;
+                }
+                sz = ht->width_bits();
+                if (member == IR::Type_Header::sizeBytes) {
+                    sz = ((sz + 7) >> 3);
+                }
+                auto result = new IR::Constant(sz);
+                result->type = IR::Type_Bits::get(32);
+                auto ctype = canonicalize(result->type);
```

Comment:
I don't think this is necessary.

Reply:
It is needed.  Without the code we get a Compiler bug.
```
hemant@ubuntu:~/int/p4c/build$ ./p4test --std p4-16 ../testdata/p4_16_samples/sizeof-bmv2.p4
SZ: 112
SZ: 112
In file: /home/hemant/int/p4c/frontends/p4/typeChecking/typeChecker.cpp:169
Compiler Bug: Could not find type of 112

hemant@ubuntu:~/int/p4c/build$ 
```

Reply:
`canonicalize` on a Type_Bits type is a noop (always returns the argument), so the call to `canonicalize is not needed.  In addition, the type could/should be set in the ctor:

    auto result = new IR::Constant(IR::Type::Bits::get(32), sz);
    setType(result, result->type);

Reply:
Though arguably, the result of one of these `size` operations should be an untyped constant (a Type_InfInt), not a bit<32>

Reply:
Thanks, Chris.   Hmm, when would a header size be greater than `2^32`?  If we go with TypeIInfInt, I will have to see what to use replace `get(32)` with.

Reply:
It might be much smaller than 2^32 -- it would be reasonable to store the result of a .size in eg a bit<8> field.  It should be possible to use just `result = IR::Constant(sz);` giving a constant of unspecified size and have the typechecker infer the size from the context.

Reply:
Soon as I use InfInt this code does not even compile.  My impression is that an InfInt is written in a P4 program as `int`.

```
struct Parsed_packet {
    Ethernet_h    ethernet;
    ipv4_t        ipv4;
    int           length;
}
```
```
hemant@ubuntu:~/int/p4c/build$ ./p4test --std p4-16 ../testdata/p4_16_samples/sizeof-bmv2.p4
../testdata/p4_16_samples/sizeof-bmv2.p4(51):syntax error, unexpected IDENTIFIER, expecting <
    int length
        ^^^^^^
error: 1 errors encountered, aborting compilation
hemant@ubuntu:~/int/p4c/build$ 
```

Reply:
It is expected that the compiler should give an error for `P4_16` code like that.  Type `int` is not allowed in headers.

It is allowed to assign a value of type `int` to an L-value of type `bit<W>` or `int<W>` in P4_16 programs, with a warning if the compile-time constant value of the `int` is too wide for the L-value being assigned to.  p4c already implements all of this, I believe.

I don't think anyone is suggesting that we should change P4_16 so that headers can contain values of type `int`.  I believe the suggested change is that `sizeBits()` should return a value of type `int`.

Reply:
Ah, I got it about int not allowed in headers.  I will what to do towards assigning the computed length to another const type.  thanks.